osu.Game/Database/RealmLive.cs

```diff
@@ -102,10 +119,23 @@ public T Value
                 if (!ThreadSafety.IsUpdateThread)
                     throw new InvalidOperationException($"Can't use {nameof(Value)} on managed objects from non-update threads");
 
-                return realmFactory.Context.Find<T>(ID);
+                ensureDataIsFromUpdateThread();
+
+                return data;
             }
         }
 
+        private void ensureDataIsFromUpdateThread()
+        {
+            Debug.Assert(ThreadSafety.IsUpdateThread);
+
+            if (dataIsFromUpdateThread)
+                return;
+
+            dataIsFromUpdateThread = true;
+            data = realmFactory.Context.Find<T>(ID);
```

Comment:
What's the deal with this refetch?

Reply:
If the live object wasn't created on the update thread, it's required to re-fetch it on the first access from the update thread. I think this may answer your other comment above too.

Alternatively, we can enforce that `ToLive` is only allowed on the update thread, but this limits its usefulness further.

Reply:
So if you somehow retrieve an object off-thread, store it, then construct a `RealmLive` from it yourself on the update thread? Is that the case guarded against here? Is there no way to do like `data.Realm == realmFactory.Context` for this case?

Reply:
```csharp
ILive<BeatmapInfo> beatmap;

private void load(RealmContextFactory contextFactory)
{
    using (var realm = contextFactory.CreateContext())
        beatmap = realm.All<BeatmapInfo>.First().ToLive;
}
```

is the simplest example i can think of

Reply:
But then the `dataIsFromUpdateThread` value set in the ctor would be `false`?

Reply:
Correct, which will cause a re-fetch on first access from update thread via the `ensure...` method.

Reply:
Nevermind I misread the code.